src/Microsoft.VisualStudio.Threading/RarelyRemoveItemSet`1.cs

```diff
@@ -0,0 +1,242 @@
+
+            this.value = null;
+            this.count = 0;
+
+            return copy;
+        }
+
+        /// <summary>
+        /// Make a thread safe copy of the content of this list.
+        /// </summary>
+        public T[] ToArray()
+        {
+            if (this.count == 0)
+            {
+                return Array.Empty<T>();
+            }
+
+            var results = new T[this.count];
+            if (this.value is T[] valueArray)
+            {
+                Array.Copy(valueArray, results, this.count);
+            }
+            else
+            {
+                results[0] = (T)this.value!;
+            }
+
+            return results;
+        }
+
+        public struct Enumerable
+        {
+            private readonly object? value;
+
+            /// <summary>
+            /// The number of items.
+            /// </summary>
+            private readonly int count;
+
+            public Enumerable(object? value, int count)
+            {
+                this.value = value;
+                this.count = count;
+            }
+
+            /// <summary>
+            /// Returns an enumerator for a current snapshot of the collection.
+            /// </summary>
+            public Enumerator GetEnumerator()
+            {
+                return new Enumerator(this.value, this.count);
+            }
+        }
+
+        public struct Enumerator : IEnumerator<T>
+        {
+            private const int IndexBeforeFirstArrayElement = -1;
+            private const int IndexSingleElement = -2;
+            private const int IndexBeforeSingleElement = -3;
+
+            private readonly object? enumeratedValue;
+            private readonly int count;
+
+            private int currentIndex;
+
+            internal Enumerator(object? enumeratedValue, int count)
+            {
+                this.enumeratedValue = enumeratedValue;
+                this.count = count;
+                this.currentIndex = 0;
+                this.Reset();
+            }
+
+            public T Current
+            {
+                get
+                {
+                    if (this.currentIndex == IndexBeforeFirstArrayElement || this.currentIndex == IndexBeforeSingleElement)
+                    {
+                        throw new InvalidOperationException();
+                    }
+
+                    // enumeratedValue cannot be null here following a call to `MoveNext` that returns true (required
+                    // for correct usage of IEnumerator).
+                    return this.currentIndex == IndexSingleElement
+                        ? (T)this.enumeratedValue!
+                        : ((T[])this.enumeratedValue!)[this.currentIndex];
+                }
+            }
+
+            object System.Collections.IEnumerator.Current
+            {
+                get { return this.Current; }
+            }
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
```

Comment:
If this is called a lot, it may be worth reducing overall branching by adding a top-level if/else on < 0.

Reply:
thanks, that is a very good point.  I copied it over, and didn't pay attention to the detail.  Yeah, it can be cleaned up to optimize for the common path.